### PR TITLE
qemu_guest_agent.py: Move qga installation code logic away

### PR DESCRIPTION
**Pull request by 6-dehan**
Move qga installation code logic to internal file,
and change some parameter name.

ID: 1890839, 1886685
Signed-off-by: Dehan Meng <demeng@redhat.com>

### CODE REVIEW

**Comment by 6-dehan** [comment]
@qizhu please have a review when you're free,thanks.

File: qemu/tests/cfg/qemu_guest_agent.cfg
Comment (vivianQizhu) [inline qemu/tests/cfg/qemu_guest_agent.cfg:46]: This is not fully automated, right? You manually put the rpm package here before the test start?

Comment (6-dehan) [reply qemu/tests/cfg/qemu_guest_agent.cfg:46]: no, the qga rpm packages were installed and downloaded locally,and these operations is executed by internal .
It's my fault , I should @you in gitlab for another part of this patch.

Comment (vivianQizhu) [reply qemu/tests/cfg/qemu_guest_agent.cfg:46]: @6-dehan If people has no your internal tool, how should they test with this script? Better give a instruction with comment and also I recommend you have the rpm location instead of the command as the variable.

Comment (vivianQizhu) [reply qemu/tests/cfg/qemu_guest_agent.cfg:46]: Recommend:
1. Name it `qga_rpm_path`.
2. Allow 2 ways to define it: a. Provide path of the  directly; b. A shell command to get it like here you have.
3. Allow it to be NULL, in this case you do nothing for the installation.

To implement 2,
`qga_rpm_path = cmd:ls /qemu-guest-agent*`
```
qga_rpm_path = params.get(qga_rpm_path')
if qga_rpm_path and qga_rpm_path.startswith('cmd:'):
qga_rpm_path = session.cmd_output_safe(qga_rpm_path.split(':')[1])
```

**Comment by 6-dehan** [comment]
@vivianQizhu 
python3 ConfigTest.py --testcase=qemu_guest_agent.virtio_serial.gagent_install.q35 --guestname=RHEL.8.3.1 --platform=x86_64 --machine=q35   --customsparams="module_stream_n_release_ver = 8.3:831"
(1/3) io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads: PASS (1427.65 s)(depends on patch: gitlab-514)
(2/3) type_specific.io-github-autotest-qemu.rh_kernel_update: PASS (137.31 s)
(3/3) type_specific.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.gagent_install: PASS (79.87 s)

**Review by vivianQizhu — APPROVED** [review]

**Comment by vivianQizhu** [comment]
@yanan-fu Request for your review, thanks.

File: qemu/tests/qemu_guest_agent.py
Comment (yanan-fu) [inline qemu/tests/qemu_guest_agent.py:191]: Why do you change to use hard code `qemu-guest-agent` here ?   The old design that use the output of `gagent_pkg_check_cmd `  is more robust i think.

Comment (yanan-fu) [reply qemu/tests/qemu_guest_agent.py:191]: with the old design, if anything changed in the future, change param `gagent_pkg_check_cmd = "rpm -q qemu-guest-agent"` is enough.

Comment (6-dehan) [reply qemu/tests/qemu_guest_agent.py:191]: okay, got it

File: qemu/tests/qemu_guest_agent.py
Comment (yanan-fu) [inline qemu/tests/qemu_guest_agent.py:164]: With this new change, there is no long the concept of the `latest`,  it is used to install the specified qga.

Comment (6-dehan) [reply qemu/tests/qemu_guest_agent.py:164]: they're moved into internal project, right? Did you check our internal project? pullrequest_514.

Comment (yanan-fu) [reply qemu/tests/qemu_guest_agent.py:164]: No, i know that. I mean the concept is not `latest` now.  It is used to install the specified qga(/qemu-guest-agent*) with the given path, it depends on the user.

Comment (6-dehan) [reply qemu/tests/qemu_guest_agent.py:164]: so you mean this function should better change func_name, because it's no longer meaning original concept right? if I got your point correctly, what do you think about naming it as "get_qga_pkg", and add a comment line ‘#’ to explain that we obtain the latest qga package through our special tool.

Comment (yanan-fu) [reply qemu/tests/qemu_guest_agent.py:164]: Yeap, that is what i mean.  And from my view, no need to comment that we use latest through other tools in upstream.

Comment (6-dehan) [reply qemu/tests/qemu_guest_agent.py:164]: okay, got it, and the name  "_get_qga_pkg" you think is okay, I'll change it

File: qemu/tests/qemu_guest_agent.py
Comment (yanan-fu) [inline qemu/tests/qemu_guest_agent.py:166]: here, still use latest

File: qemu/tests/qemu_guest_agent.py
Comment (yanan-fu) [inline qemu/tests/qemu_guest_agent.py:326]: same , still use latest

Comment (6-dehan) [reply qemu/tests/qemu_guest_agent.py]: okay, my bad, updated again.

File: qemu/tests/qemu_guest_agent.py
Comment (yanan-fu) [inline qemu/tests/qemu_guest_agent.py:127]: Please go through the whole context, the param name already changed.

And, be careful with the document, such as L115 and so on.

Comment (6-dehan) [reply qemu/tests/qemu_guest_agent.py:127]: Changed "latest" as "specific" this time.

**Comment by 6-dehan** [comment]
@yanan-fu , could you have a check ? thanks.

**Comment by 6-dehan** [comment]
@yanan-fu , Because I run a test loop today and hit related issues for test case "qemu_guest_agent_hotplug", debugged it and edited a new .cfg file as sync, could you have a check ? thanks.
(1/4) Host_RHEL.m8.u4.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads: PASS (1670.21 s)
(2/4) Host_RHEL.m8.u4.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.rh_kernel_update: PASS (123.59 s)
(3/4) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.gagent_install: PASS (103.29 s)
(4/4) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.qemu_guest_agent_hotplug: ERROR: QMP command 'chardev-remove' failed    (arguments: {'id': 'qga-serial0'},    error message: {'class': 'GenericError', 'desc': "Chardev 'qga-serial0' is busy"}) (50.95 s) (because of another product bug, anyway our aim part was passed,  so ignore it please.)

**Comment by yanan-fu** [comment]
@6-dehan  Please check the comment in the issue, thanks!

**Comment by 6-dehan** [comment]
@yanan-fu could you please have a check with this new code? 
(1/3) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (1571.16 s)
 (2/3) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.rh_kernel_update.q35: PASS (118.86 s)
 (3/3) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.gagent_install.q35: PASS (99.51 s)

**Comment by 6-dehan** [comment]
@yfu The newest code has been updated, please have a review with it, thanks. Here are the results: RHEL840_guest on RHEL850_host: (1/3) Host_RHEL.m8.u5.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (965.72 s) (2/3) Host_RHEL.m8.u5.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.rh_kernel_update.q35: PASS (94.63 s) (3/3) Host_RHEL.m8.u5.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.gagent_install.q35: PASS (56.01 s)

RHEL840_guest on RHEL900_host: (1/3) Host_RHEL.m9.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (1216.66 s) (2/3) Host_RHEL.m9.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.rh_kernel_update.q35: PASS (123.15 s) (3/3) Host_RHEL.m9.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.gagent_install.q35: PASS (96.01 s)

RHEL831_guest on RHEL900_host: (1/3) Host_RHEL.m9.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.1.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (1672.28 s) (2/3) Host_RHEL.m9.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.1.x86_64.io-github-autotest-qemu.rh_kernel_update.q35: PASS (173.12 s) (3/3) Host_RHEL.m9.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.1.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.gagent_install.q35: PASS (100.41 s)

RHEL830_guest on RHEL850_host: (1/3) Host_RHEL.m8.u5.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (1504.60 s) (2/3) Host_RHEL.m8.u5.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.rh_kernel_update.q35: PASS (106.49 s) (3/3) Host_RHEL.m8.u5.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.gagent_install.q35: PASS (59.13 s)

RHEL811_guest on RHEL900_host: (1/3) Host_RHEL.m9.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.1.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (1699.27 s) (2/3) Host_RHEL.m9.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.1.x86_64.io-github-autotest-qemu.rh_kernel_update.q35: PASS (328.41 s) (3/3) Host_RHEL.m9.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.1.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.gagent_install.q35: PASS (103.95 s)

**Comment by 6-dehan** [comment]
@yfu, thanks for your advice.
(1/3) Host_RHEL.m8.u5.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (1467.19 s)
 (2/3) Host_RHEL.m8.u5.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.rh_kernel_update.q35: PASS (101.04 s)
 (3/3) Host_RHEL.m8.u5.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.gagent_install.q35: PASS (60.08 s)

File: qemu/tests/cfg/qemu_guest_agent.cfg
Comment (yanan-fu) [inline qemu/tests/cfg/qemu_guest_agent.cfg:21]: Please update the file name.

Comment (6-dehan) [reply qemu/tests/cfg/qemu_guest_agent.cfg:21]: @yanan-fu, done.

Comment (yanan-fu) [reply qemu/tests/cfg/qemu_guest_agent.cfg:21]: @6-dehan  Seems no change, could you double confirm ? thanks!

Comment (6-dehan) [reply qemu/tests/cfg/qemu_guest_agent.cfg]: @yanan-fu  sorry about that. something went wrong at that time. It's okay now.

**Review by yanan-fu — APPROVED** [review]
Acked-by: Yanan Fu <yfu@redhat.com>

**Comment by yanan-fu** [comment]
@vivianQizhu Could you help review the latest change ? Thanks!

File: qemu/tests/qemu_guest_agent.py
Comment (vivianQizhu) [inline qemu/tests/qemu_guest_agent.py:139]: There is a risk, `version_list[1]` might hit index error since`build_specific` could be `None` if nothing located at `self.qga_pkg_path`(for other users who might not place it in advance).

Comment (6-dehan) [reply qemu/tests/qemu_guest_agent.py]: @vivianQizhu thanks for your suggestion, I add a if statement to check whether ```version_list[1]``` is none. and provide a error message for other users. Could you please have a check?

File: qemu/tests/qemu_guest_agent.py
Comment (6-dehan) [reply qemu/tests/qemu_guest_agent.py:144]: @vivianQizhu , thanks. Code has been updated.

File: qemu/tests/qemu_guest_agent.py
Comment (vivianQizhu) [reply qemu/tests/qemu_guest_agent.py:136]: @6-dehan What do you think about this one? If you can not get `full_qga_rpm_info`, all the later actions could be skipped.

Comment (6-dehan) [reply qemu/tests/qemu_guest_agent.py:136]: @vivianQizhu , More reasonable for our code logic, okay, I'll update right now. thanks.

**Review by vivianQizhu — APPROVED** [review]

**Comment by 6-dehan** [comment]
@yanan-fu @vivianQizhu results updated:
RHEL850-AV
(1/2) Host_RHEL.m8.u5.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.5.0.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (958.76 s)
 (2/2) Host_RHEL.m8.u5.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.5.0.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.gagent_install.q35: PASS (86.68 s)
RHEL850-SLOW
(01/50) Host_RHEL.m8.u5.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.5.0.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (961.76 s)
 (02/50) Host_RHEL.m8.u5.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.5.0.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.gagent_install.q35: PASS (60.59 s)

**Comment by vivianQizhu** [comment]
Thanks @6-dehan and @yanan-fu for all your effort. @yanan-fu Would you help double check the latest version before merging it, thanks.

File: qemu/tests/qemu_guest_agent.py
Comment (yanan-fu) [inline qemu/tests/qemu_guest_agent.py:139]: It should output the the version info after the end of the loop.

File: qemu/tests/qemu_guest_agent.py
Comment (yanan-fu) [reply qemu/tests/qemu_guest_agent.py]: I have a bit different opinion about this.

If can not get version info with the pattern, i prefer to use '', instead of append nothing.   Other wise, it looks puzzled for the following code if there len is 1.
```
logging.info("The installed and the specific pkg "
                   "version is %s", version_list)
```

By the way, it is not absolute that `len(version_list) < 2` is caused by no proper qga be specified, since no installed qga can turn to this result too.

So i suggest use `` when failed to get the version with pattern, and compare version_list[1] and  version_list[0].

Comment (vivianQizhu) [reply qemu/tests/qemu_guest_agent.py]: If no installed qga `s, o = session.cmd_status_output(cmd_check_pkg)` `s == 0` will be `False` then it will not step into here.
And if you assume there is exception that it might not be installed, then it is possible that they are both `""` and then the function will return `True`, is that expected?
However, both okay for me if the logic is correct and expected and also the code is not in low efficiency.

Comment (vivianQizhu) [reply qemu/tests/qemu_guest_agent.py]: If no installed qga `s, o = session.cmd_status_output(cmd_check_pkg)` `s == 0` will be `False` then it will not step into here.
And if you assume there is exception that it might not be installed, then it is possible that they are both `""` and then the function will return `True`, is that expected?
However, both okay for me if the logic is correct and expected and also the code is not in low efficiency.

Comment (yanan-fu) [reply qemu/tests/qemu_guest_agent.py]: > If no installed qga `s, o = session.cmd_status_output(cmd_check_pkg)` `s == 0` will be `False` then it will not step into here.

Yes, you are right, in this case it will return False directly, this is the expected result. I miss this before.

> And if you assume there is exception that it might not be installed, then it is possible that they are both `""` and then the function will return `True`, is that expected?

Not the expected, and the explanation above can avoid this, i just do not want to see the debug info like the following example when no proper qga package be provided.

`The installed and the specific pkg version is [$INSTALLED_PKG]`

> However, both okay for me if the logic is correct and expected and also the code is not in low efficiency.

Comment (yanan-fu) [reply qemu/tests/qemu_guest_agent.py]: Eh~  Get the installed qga pkg is not equal to can get install qga version, so we still can not avoid both `""`.
Bases on this, i am ok with this logic ~

Comment (vivianQizhu) [reply qemu/tests/qemu_guest_agent.py]: @yanan-fu If move the `logging` line above to `qga_v = re.findall(pattern, pkg, re.I)`, anyway when it steps there both the installed one and `full_qga_rpm_info` are all exist, simply log it like `"Found qga pkg: %s" % pkg`, what do you think?

Comment (yanan-fu) [reply qemu/tests/qemu_guest_agent.py]: @vivianQizhu In this way, it can not distinguish the installed pkg and specified pkg from the log. No friendly for debugging, do you think this is a problem ? Eh~   not a big issue :)

Or we can use dict instead of list to save the version info ?

Comment (vivianQizhu) [reply qemu/tests/qemu_guest_agent.py]: @yanan-fu I have thought about dict too, but yes not a big deal, so let's leave it to @6-dehan at his discretion.

**Comment by 6-dehan** [comment]
@vivianQizhu @yanan-fu , thanks for your two comments, I went through those, and actually, the latest demand is to output qga_ver Separately and clearly, am I right? and you two suggested changing List to Dict. 
For me side, the original code logic has sorted the two in '[" ", " "]', that is, the former is installed by default, and the latter is the specified version. As Yanan said: Successfully obtaining the installed rpm does not mean that the qga_ver can be successfully obtained. If this occurs,  the problem is mainly the 'pattern' problem that there is no precise matching. If the tag of rpm changes in the future, it can be targeted and optimized, right?  so if this issue is not a big deal, leave it here now it's okay from my side. maybe I could optimize it later.

**Comment by yanan-fu** [comment]
> @vivianQizhu @yanan-fu , thanks for your two comments, I went through those, and actually, the latest demand is to output qga_ver Separately and clearly, am I right? and you two suggested changing List to Dict.
> For my side, the original code logic has sorted the two in '[" ", " "]', that is, the former is installed by default, and the latter is the specified version. As Yanan said: Successfully obtaining the installed rpm does not mean that the qga_ver can be successfully obtained. If this occurs, the problem is mainly the 'pattern' problem that there is no precise matching. If the tag of rpm changes in the future, it can be targeted and optimized, right? so if this issue is not a big deal, leave it here now it's okay from my side. maybe I could optimize it later.
> 
> if I misunderstood anything, please point it out. thanks

I am ok with that, thanks!

File: qemu/tests/qemu_guest_agent.py
Comment (vivianQizhu) [reply qemu/tests/qemu_guest_agent.py:139]: Despite the dict and list issue, I think it should at least be moved out from the `for` loop. Now the info will be print twice and the first round is not correct.

Comment (yanan-fu) [reply qemu/tests/qemu_guest_agent.py:139]: Yes~   This is not controversial ~   @6-dehan

Comment (6-dehan) [reply qemu/tests/qemu_guest_agent.py]: > Despite the dict and list issue, I think it should at least be moved out from the `for` loop. Now the info will be print twice and the first round is not correct.

@vivianQizhu @yanan-fu the logging.info has change out of the `for` loog, here are result of new code. I checked the log that the qga_v info print once as expected.
(01/50) Host_RHEL.m8.u5.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.5.0.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (937.98 s)
 (02/50) Host_RHEL.m8.u5.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.5.0.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.gagent_install.q35: PASS (95.89 s)

**Comment by vivianQizhu** [comment]
Thanks @6-dehan and @yanan-fu

**Pull request merged**
